tokio/src/sync/broadcast.rs

```diff
@@ -601,6 +608,9 @@ impl<T> Sender<T> {
         // Release the slot lock
         slot.lock.store(0, SeqCst);
 
+        // Release the mutex
+        drop(tail);
```

Comment:
why is the `tail` mutex  held to this point?

Reply:
Another thread could send on the same slot. If tail is dropped before `slot.lock.store(0, SeqCst);` then it could clear the writer bit while the other thread is in the critical section.

Reply:
I see, could you add a comment explaining that as well?

Given that this fixes a bug, we probably should move forward w/ this. It is unfortunate that the scope of the mutex is so big, but we can worry about it later.

Reply:
Yes, it's unfortunate. Another way this could be solved would be using a semaphore to limit the number of writers in send to `self.shared.buffer.len()`. In that case, the other changes to `send2` wouldn't be needed.

Reply:
The semaphore strategy is probably a better option in general and *should* simplify the implementation.

That said, we don't have any benchmarks to measure performance right now, so we should focus on correctness. I'm fine w/ this change now, and would always appreciate follow up work to add benches & cleanup the impl potentially adding a semaphore as you mentioned :)

Reply:
Done.